runtime/lua/vim/snippet.lua

```diff
@@ -0,0 +1,212 @@
+local snippet = require('vim.lsp._snippet')
+
+---@type { session: nvim.SnippetSession }
+local state = {}
+
+---Return one indent string of current buffer.
+---@return string
+local function get_one_indent()
+  local expandtab = vim.api.nvim_get_option('expandtab')
+  if not expandtab then
+    return '\t'
+  end
+  local shiftwidth = vim.api.nvim_get_option('shiftwidth')
+  if shiftwidth > 0 then
+    return string.rep(' ', shiftwidth)
+  end
+  return string.rep(' ', vim.api.nvim_get_option('tabstop'))
+end
+
+---Return base indent of current line.
+---@return string
+local function get_base_indent()
+  return string.match(vim.api.nvim_get_current_line(), '^%s*') or ''
+end
+
+---Select specific range.
+---@param range { s: { number, number }, e: { number, number } }
+local function select_text(range)
+  if range.s[1] == range.e[1] and range.s[2] == range.e[2] then
+    vim.api.nvim_win_set_cursor(0, { range.s[1] + 1, range.s[2] })
+  else
+    vim.api.nvim_feedkeys(vim.api.nvim_replace_termcodes([[<Esc>%sG%s|v%sG%s|<C-g>]], true, true, true):format(
+      range.s[1] + 1,
+      range.s[2] + 1,
+      range.e[1] + 1,
+      range.e[2]
+    ), 'in', true)
+  end
+end
+
+---Adjust snippet indent
+---@param snippet_text string
+---@return string
+local function adjust_indent(snippet_text)
+  local one_indent = get_one_indent()
+  local base_indent = get_base_indent()
+  local adjusted_snippet_text = {}
+  local texts = vim.split(snippet_text, '\n', true)
+  for i, text in ipairs(texts) do
+    if i ~= #texts and string.match(text, '^%s*$') then
+      table.insert(adjusted_snippet_text, '')
+    else
+      if i ~= 1 then
+        text = string.gsub(text, '^(\t+)', function(indent)
+          return string.gsub(indent, '\t', one_indent)
+        end)
+        table.insert(adjusted_snippet_text, base_indent .. text)
+      else
+        table.insert(adjusted_snippet_text, text)
+      end
+    end
+  end
+  return table.concat(adjusted_snippet_text, '\n')
+end
+
+---@alias nvim.SnippetJumpDirection "1" | "2"
+local JumpDirection = {}
+JumpDirection.Next = 1
+JumpDirection.Prev = 2
+
+
+---@class nvim.SnippetSession
+---@field ns string
+---@field public snippet_text string
+local SnippetSession = {}
+
+---Create SnippetSession instance.
+---@param snippet_text string
+function SnippetSession.new(snippet_text)
+  return setmetatable({
+    ns = vim.api.nvim_create_namespace('nvim.snippet'),
+    snippet_text = adjust_indent(snippet_text),
+  }, { __index = SnippetSession })
+end
+
+---Insert snippet to the current buffer.
+function SnippetSession:expand()
+  local cursor = vim.api.nvim_win_get_cursor(0)
+  cursor[1] = cursor[1] - 1
+
+  local snippet_node = snippet.parse(self.snippet_text)
+  vim.api.nvim_buf_set_text(0, cursor[1], cursor[2], cursor[1], cursor[2], vim.split(tostring(snippet_node), '\n'))
```

Comment:
Currently, this will insert the result of `tostring(snippet_ast)`.
It should be changed when implementing LSP variables.